src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+            if ((operations[0] as Operation)?._parentDoNotAccessDirectly != null)
+            {
+                // already initialized
+                return operations;
+            }
+
+            foreach (var operation in operations)
+            {
+                // go through slowest path
+                SetParentOperation(operation, parent);
             }
 
-            public OperationKind Kind => OperationKind.None;
+            return operations;
+        }
 
-            public SyntaxNode Syntax { get; }
+        public static T ResetParentOperation<T>(T operation) where T : IOperation
+        {
+            if (operation == null)
+            {
+                return operation;
+            }
+
+            Interlocked.Exchange(ref (operation as Operation)._parentDoNotAccessDirectly, null);
+            return operation;
+        }
 
-            public ITypeSymbol Type => null;
+        private class NoneOperation : Operation
+        {
+            private readonly Func<ImmutableArray<IOperation>> _getChildren;
 
-            public Optional<object> ConstantValue { get; }
+            public NoneOperation(SemanticModel semanticMode, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren) :
+                base(OperationKind.None, semanticMode, node, type: null, constantValue: constantValue)
+            {
+                _getChildren = getChildren;
+            }
 
-            public void Accept(OperationVisitor visitor)
+            public override void Accept(OperationVisitor visitor)
             {
                 visitor.VisitNoneOperation(this);
             }
 
-            public TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
+            public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
             {
                 return visitor.VisitNoneOperation(this, argument);
             }
 
-            public IEnumerable<IOperation> Children => _getChildren().NullToEmpty();
+            public override IEnumerable<IOperation> Children
+            {
+                get
+                {
+                    foreach (var child in _getChildren().NullToEmpty().WhereNotNull())
+                    {
+                        yield return Operation.SetParentOperation(child, this);
+                    }
+                }
+            }
+        }
+
+        private static readonly ObjectPool<Queue<IOperation>> s_queuePool =
+            new ObjectPool<Queue<IOperation>>(() => new Queue<IOperation>(), 10);
+
+        private IOperation WalkDownOperationToFindParent(HashSet<IOperation> operationAlreadyProcessed, IOperation root)
+        {
+            void EnqueueChildOperations(Queue<IOperation> queue, IOperation parent)
+            {
+                // children can return null
+                foreach (var o in parent.Children.WhereNotNull())
```

Comment:
currently Children can return null. some because bound node is giving ImmutableArray with null children for invalid expression. some because how children property is implemented.

we need to update children to filter out null child, but that will be another PR.